src/topology/metric_space/path_metric.lean

```diff
@@ -0,0 +1,225 @@
+    { ext, refl }, norm_num },
+  all_goals { apply ((set.monotone_proj_Icc _).comp $ λ x y h, _).monotone_on },
+  { exact mul_le_mul_of_nonneg_left h zero_le_two },
+  { exact sub_le_sub_right (mul_le_mul_of_nonneg_left h zero_le_two) _ },
+end
+
+lemma comp_length_le {F : Type*} [pseudo_emetric_space F] {φ : E → F} {K : ℝ≥0}
+  (hφ : lipschitz_with K φ) : (p.map hφ.continuous).length ≤ ↑K * p.length :=
+lipschitz_on_with.comp_evariation_on_le (hφ.lipschitz_on_with set.univ) (set.maps_to_univ _ _)
+
+/-- An alternative definition of `path.length`. -/
+lemma evariation_on_extend_unit_interval_eq_length :
+  evariation_on p.extend I = p.length := arclength_Icc_extend zero_le_one p
+
+lemma arclength_on_zero_one_eq_length : arclength p 0 1 = p.length :=
+by rw [length, arclength, unit_interval.univ_eq_Icc]
+
+lemma length_of_continuous_on {X : Type*} {f : ℝ → X} {s t : ℝ} (hst : s ≤ t)
+  [pseudo_emetric_space X] (hf : continuous_on f (set.Icc s t)) :
+  (of_continuous_on hst hf).length = arclength f s t :=
+begin
+  apply evariation_on.comp_eq_of_monotone_on _ _ (monotone.monotone_on _ _) _ (λ x hx, _),
+  { exact (monotone_affine_of_le hst).comp (λ _ _, id) },
+  { exact λ x hx, affine_maps_to_I hst x.2 },
+  { obtain ⟨y, hy, h'⟩ := affine_surj_on_I hx, exact ⟨⟨y, hy⟩, trivial, h'⟩ },
+end
+
+end path
+
+section path_emetric
+
+/-- The type synonym for the pseudo-emetric space `E` endowed with its path pseudo-emetric. -/
+def path_emetric (E : Type*) := E
+
+variables {E : Type*}
+
+instance [h : inhabited E] : inhabited (path_emetric E) := ⟨h.1⟩
+
+/-- Casting from `E` to `path_emetric E`. -/
+def to_path_emetric : E ≃ path_emetric E := equiv.refl _
+
+/-- Casting from `path_emetric E` to `E`. -/
+def from_path_emetric : path_emetric E ≃ E := equiv.refl _
+
+lemma from_to_path_emetric (x : E) : from_path_emetric (to_path_emetric x) = x := rfl
+lemma to_from_path_emetric (x : path_emetric E) : to_path_emetric (from_path_emetric x) = x := rfl
+
+local notation `of` := to_path_emetric
+local notation `fo` := from_path_emetric
+
+variables [pseudo_emetric_space E]
+
+/--
+The path pseudo-emetric on `path_emetric E`:
+The distance between two points is the infimum of the length of paths joining them
+-/
+instance : pseudo_emetric_space (path_emetric E) :=
+{ edist := λ x y, ⨅ p : path (fo x) (fo y), p.length,
+  edist_self := λ x, le_antisymm (infi_le_of_le _ $ (path.length_refl _).le) zero_le',
+  edist_comm := λ x y, by apply le_antisymm;
+    exact le_infi (λ p, (infi_le _ _).trans_eq p.length_symm),
+  edist_triangle := λ x y z, begin
+    simp_rw [ennreal.infi_add, ennreal.add_infi],
+    exact le_infi₂ (λ p q, (infi_le _ _).trans_eq $ p.length_trans q),
+  end }
+
+/-- `E` is a length space if the path emetric on `E` is equal to its original metric. -/
+class length_space (E) [pseudo_emetric_space E] : Prop :=
+(isom : isometry (of : E → path_emetric E))
```

Comment:
Should we add `geodesic_space` as well?

Reply:
Shouldn't that get its own file?

Reply:
I agree, that should be a separate file (or at the very least, a separate PR).